### PR TITLE
Sync project

### PR DESCRIPTION
**Pull request by rob-baron**
This is to sync atmosphere projects in such a way that they act like openstack projects.  This does not sync instances within these projects, that is a future pull request.

### CODE REVIEW

**Comment by xuhang57** [comment]
each PR should really focus on one feature. This PR contains multiple and it would be difficult for other people to test it.

**Comment by jeremyfreudberg** [comment]
Agree with Lucas here. Also seems like each PR that you've opened here contains overlapping commits which can get confusing.

**Comment by rob-baron** [comment]
This one is dependent on the MOC Login without Project name.

Not quite sure how I can make it independent of that one.

I am breaking things down to vertical slices of functionality.  The first one is that the user can log in using only their username and password.  This one is that users can view their open stack projects after they log on.

**Comment by jeremyfreudberg** [comment]
There's no good workflow that I can think of to truly make your patches independent. We either have to do code review faster, or you should just have one PR for everything (but which contains multiple, clear commits)

**Comment by rob-baron** [comment]
So SyncProject contains the changes that MocOpenstackLogin has.  I have restructured it slightly from my starting point in order to be PEP8 compliant and to be a little more rational with open stack.  Celery is a pain to deal with, however it is needed.

I am expecting to do a squash commit for this part, as I have had a number of small commits.  Note: I didn't change the README.md file for this commit.

**Comment by rob-baron** [comment]
I have squashed commits made from the start of this project.

File: README.md [hunk 1]
```diff
@@ -62,14 +62,11 @@ Imports should be grouped into the sections below and in sorted order.
 
 See LICENSE.txt for license information
 
-## Lead
 
-+ **Edwin Skidmore <edwin@iplantcollaborative.org>**
-
-## Authors 
+## Original Authors
```
Comment (rob-baron) [inline README.md]: I think this file was modified by lucas and I am somehow picking up his changes.  This file is not needed for this pull request and I am not certain how to get rid of it.

**Comment by xuhang57** [comment]
Will pull and test the code soon. I tried to login as `friday-test` in the `128.31.22.8`. Login works but I don't see any image that I could use? Might add some default images?

**Review by xuhang57 — CHANGES_REQUESTED** [review]
Having CI fails and merge conflicts. Once these got resolved, I will pull and test the code. We need to figure out a way to do unit test, functional test on the backend. So other people can test the code.

**Comment by xuhang57** [comment]
In `service/tasks/driver.py`, each of these tasks is designed for the "launching instances" and "instance actions". But `Sync Project` actually belongs to the "login" process/logic. In other words, It might be ok to use celery to sync projects between OpenStack and GIJI. However, the code should probably in another place. Otherwise, it is not consistent with the code base we have now.

**Comment by rob-baron** [comment]
In the tasks files are all of the interactions with the cloud drivers that are done using celery.  These are done using celery as they are potentially long tasks and if done in line with the web, may cause the browser to time out.  Currently with atmosphere, this is just dealing with instances (as atmosphere projects are not openstack projects).

And, no, SyncProject does not belong strictly to the login process as it needs to be run periodically in order to update the project and instances.  This is required as we want changes that appear in horizon to be viewed in atmosphere.  I currently have it starting once from the login.

The plan for this is to get the projects with instances synced at login (a vertical slice of functionality), then to turn on celery's periodic running feature.

**Comment by xuhang57** [comment]
Yes, but you are not using the cloud driver, which means you should not put your code in the `driver.py`. I would recommend to create a new file and place your code there so that it is easier for other to understand. And I am not saying you should not use celery. However, does celery help you reduce the processing time? Celery makes tasks done in parallel. It would be great if you have many tasks. For `Sync Project`, you might have less than three tasks. So I don't know how much time you could reduce. BUT, using celery is fine. 

If that's the plan, you might want to call this PR: `Sync Project and Instance`. It sounds good to me to "sync project", but I am not sure about the "sync instance" part. GIJI user should not use the Horizon at all. In other words, they have no instance to sync with Horizon.

And I don't understand how celery could run periodically by itself. You have to put the tasks on the queue so these tasks would be executed. Mind elaborating this part a bit more?

**Comment by rob-baron** [comment]
1) I can move it to its own file, though I feel that it groups really well there as there are all of the interactions with celery to the open stack instance in one file.

2) The way SyncProject is laid out is that it syncs all projects.  Furthermore, there is another part that I have indicated where it will go to sync the instances.  I was planning to have another PR just for the instance part.

3) Celery is being used as SynProject calls the openstack drivers separately.  This incurs enough overhead that I am worried that if it is not placed in the background, the web browser will time out.  Furthermore, it is easily parallelized.  This is the call to get the projects can be done in the background and calls to update the instances associated with those projects can be done in parallel.

4) My reading of the Celery docs suggests that there is the provision for running the job every 5 minutes.  If you consider the following:
    @task(name="wait_for_instance", max_retries=250, default_retry_delay=15)
Here the task is retired 250 files with a delay of 15[seconds] until it succeeds.  It would be trivial to piggy back on this.  However, there is also a link parameter which links the next run after successful completion of the task.  The link parameter could be used to run the same task 5 minutes in the future - until the task fails due to a expired token.  We could also programmatically link the next run using similar code that you just eliminated in the resize.

**Comment by jeremyfreudberg** [comment]
@xuhang57 
> but I am not sure about the "sync instance" part. GIJI user should not use the Horizon at all. In other words, they have no instance to sync with Horizon.  

Shouldn't the goal be that each UI always show all instances belonging to the user's project(s), regardless of where the user launched them from? There shouldn't be anything special about an instance launched from Horizon, or from Atmosphere.  

In other words, we aren't syncing the instances with Horizon. Rather, we are making sure that Atmosphere knows about every instance that OpenStack says belongs to the user.

**Comment by xuhang57** [comment]
@jeremyfreudberg  That's fine with me. But I thought the goal here is having two GUI for different target audience. So, if you are comfortable enough to use Horizon, then you stay with it. Otherwise, use GIJI. 

This "Sync" is actually one way, since Horizon knows whatever you launch in GIJI. In other words, Horizon is still the "master" and it is for advanced users/hackers.

**Comment by xuhang57** [comment]
1. You just said you will have another PR for the instance part. So what is in this PR is for the project part. And yes, that file is about instance so that's why I am suggesting we shouldn't mix it and make it complexed. 

2. sounds great. 

3. Not sure have you tried? I don't think it would take that much time though. But yeah, what you have now, works. 

4. That task is like that due to its nature. And it would stop if an error shows up. I am not sure if it is a good idea to keep repeating the task of "sync project/instance" in the backend. It might slows down the web app.

**Comment by rob-baron** [comment]
Let's define why I called it Sync.  We are Syncing what is in atmosphere with what is in open stack.  Openstack is our reference copy.   Both troposphere/atmophere and Horizon provide views on the current state of open stack.

**Comment by rob-baron** [comment]
Lucas: #3, when I was doing the SyncProject inline it was taking an unreasonable amount of time as part of the login process (because of the calls to open stack).  When I moved it into Celery, the responsiveness of the UI increased.

**Comment by rob-baron** [comment]
Lucas: #1, No, I disagree with you.  It is a slight increase in complexity as it is a celery task that interacts with open stack.  By the atmosphere design this just happens to be all of the instance operations as atmosphere projects are not brought in.

The means that we must import the same set of libraries in a separate interpretation unit.  That said, I really don't care where we put this just so it gets done.

**Comment by rob-baron** [comment]
Lucas: #4: I was just highlighting how we specify a job in celery.  We could use celery's link to link the 2 runs together.  Perhaps simpler is to just have a loop that breaks when and error condition happens.  So the choice above is whether to contort celery to what we want (i.e. run the task x number of times once every 5 minutes until it succeeds (in our case fails)) or to consume more process space by polling (i.e. having a infinite loop which sleeps 5 minutes).  I would rather have celery do the work at it would consume fewer resources (either contorting celery or using the link feature).

**Comment by xuhang57** [comment]
1. You have to care. That said, you cannot ignore the coding style. And quoting from you: 

> I can move it to its own file, though I feel that it groups really well there as there are all of the interactions with celery to the open stack instance in one file.

so I replied based on this previously. And to reply what you said in the last sentence. Why not put all code into one file? so you don't have to import the same libs/packages. So it doesn't make sense. The reason why I would like to see them in two files is it is easier for other people to know what thoese functions about and it is easier for us to debug in the future.

**Comment by xuhang57** [comment]
I don't think we need to sync it in GIJI from OpenStack/Horizon since it is not our use case. Our users of GIJI won't go to Horizon and launch an instance and go back to GIJI. But they would launch an instance in GIJI and go to Horizon do other things. There is nothing to do with the definititon of "Sync".

**Comment by rob-baron** [comment]
lucas: #1, I can come up with argument for doing it either way.  I choose one.  

I am not talking about placing all of the code in one file, I am just saying that it clusters nicely where I placed it based on the libraries that it requires and its functionality.  Stylistically, I could go either way and still be good.  Also, complexity isn't really the issue either.

I have worked places that have as part of their style guide that every function needs to be in it's own file.  Should we implement this here?

**Comment by rob-baron** [comment]
I am closing this branch as it has been superseded with MOC_Sync_Projects.

**Pull request closed without merging**